core/blockchain.go

```diff
@@ -3015,3 +3034,155 @@ func EnablePersistDiff(limit uint64) BlockChainOption {
 		return chain
 	}
 }
+
+func (bc *BlockChain) GetRootByDiffHash(blockNumber uint64, blockHash common.Hash, diffHash common.Hash) (*types.VerifyResult, error) {
+	var res types.VerifyResult
+	res.BlockNumber = blockNumber
+	res.BlockHash = blockHash
+
+	if blockNumber > bc.CurrentHeader().Number.Uint64()+11 {
+		res.Status = types.StatusBlockTooNew
+		return &res, nil
+	} else if blockNumber > bc.CurrentHeader().Number.Uint64() {
+		res.Status = types.StatusBlockNewer
+		return &res, nil
+	}
+
+	diff := bc.GetTrustedDiffLayer(blockHash)
+	if diff != nil {
+		if diff.DiffHash == (common.Hash{}) {
+			hash, err := GetTrustedDiffHash(diff)
+			if err != nil {
+				res.Status = types.StatusUnexpectedError
+				return &res, err
+			}
+
+			diff.DiffHash = hash
+		}
+
+		if diffHash != diff.DiffHash {
+			res.Status = types.StatusDiffHashMismatch
+			return &res, nil
+		}
+
+		header := bc.GetHeaderByHash(blockHash)
+		if header == nil {
+			res.Status = types.StatusUnexpectedError
+			return &res, fmt.Errorf("unexpected error, header not found")
+		}
+		res.Status = types.StatusFullVerified
+		res.Root = header.Root
+		return &res, nil
+	}
+
+	header := bc.GetHeaderByHash(blockHash)
+	if header == nil {
+		if blockNumber > bc.CurrentHeader().Number.Uint64()-11 {
+			res.Status = types.StatusPossibleFork
+			return &res, nil
+		}
+
+		res.Status = types.StatusImpossibleFork
+		return &res, nil
+	}
+
+	res.Status = types.StatusUntrustedVerified
```

Comment:
`StatusUntrustedVerified` may be a bit misleading, what about`StatusPartiallyVerified`?

Reply:
Good idea.